### DISCUSSION TITLE
Integrating Voikko to an existing application with some custom vocabulary

### DISCUSSION THREAD

**Discussion by f4irline** [voikko/discussion#3] [Q&A] [answered]
Hey!

I have a use case where I am developing a Java application and it'd benefit from a spell checking implementation. The intention is to support at least Finnish and that's why I stumbled upon Voikko.

Everything else is fairly clear, the spell checking can be implemented by using the `libvoikko` Java API along with the `libvoikko` libraries. What is not clear to me, is that is it possible to extend the existing dictionary with some user defined words fairly easily? My current understanding is that it'd require roughly the following:

- Apache HTTP server to serve the Joukahainen web API
- PostgreSQL which would serve as a data storage for the original Joukahainen dictionary and the custom vocabulary

Then the user added words would be added to the `raw_word` table, and `joukahainen.xml.gz` (and `joukahainen.xml`) could be generated by calling the `/query/wlist` endpoint and the user added words would be included in that. Then the translations would be generated using the `make vvfst` command defined in the `corevoikko/voikko-fi` package. Is this the only way?

**Comment by hatapitk** [comment, marked as answer]
Hi! You can edit voikko-fi/vocabulary/joukahainen.xml directly and rebuild voikko-fi. No need for your own installation of Joukahainen or PostgreSQL. Not sure if that qualifies as easy but I  guess that is the easiest solution for small custom additions.

**Reply by f4irline** [reply]
Sounds simple enough! Is the format documented somehow somewhere? Is it possible to create very simple "dummy" additions, which would just not show up as errors and would show up in the suggestions? So what are the absolutely minimum _required_ fields 🙂

**Reply by hatapitk** [reply]
https://joukahainen.puimula.org/docs/index.html has some documentation but it does not say which elements in the XML format correspond to which fields in Joukahainen - usually that is fairly easy to guess by looking at some examples from the vocabulary.

Anyway if you know Finnish enough to be able to find a word that matches the part of speech and inflection class of the word that you want to add, just find that word from XML, make a copy of it and replace the content of 'form' element. This allows the word to be inflected and used in compounds and does not break grammar checking.

If you do not know any Finnish, use `<wclass>interjection</wclass>` and `<form>yourwordhere</form>`, leaving other elements out. The word also needs an 'id' but that can be anything. This allows the word to be accepted in spell checking in that exact form but may cause problems with grammar checking or morphological analysis should you need those.
